data/src/main/java/org/akvo/flow/data/datasource/DatabaseDataSource.java

```diff
@@ -0,0 +1,184 @@
+
+    @Inject
+    public DatabaseDataSource(BriteDatabase db) {
+        this.briteSurveyDbAdapter = new BriteSurveyDbAdapter(db);
+    }
+
+    public Observable<Cursor> getDataPoints(@NonNull Long surveyGroupId, @Nullable Double latitude,
+            @Nullable Double longitude, @Nullable Integer orderBy) {
+        if (isRequestFiltered(orderBy)) {
+            return briteSurveyDbAdapter
+                    .getFilteredDataPoints(surveyGroupId, latitude, longitude, orderBy);
+        } else {
+            return briteSurveyDbAdapter.getDataPoints(surveyGroupId);
+        }
+
+    }
+
+    private boolean isRequestFiltered(@Nullable Integer orderBy) {
+        return orderBy != null && (orderBy == Constants.ORDER_BY_DISTANCE ||
+                orderBy ==  Constants.ORDER_BY_DATE ||
+                orderBy ==  Constants.ORDER_BY_STATUS ||
+                orderBy ==  Constants.ORDER_BY_NAME);
+        return false;
+    }
+
+    public Cursor getSyncedTime(long surveyGroupId) {
+        return briteSurveyDbAdapter.getSyncTime(surveyGroupId);
+    }
+
+    public Observable<List<ApiDataPoint>> syncDataPoints(List<ApiDataPoint> apiDataPoints) {
+        if (apiDataPoints == null) {
+            return Observable.<List<ApiDataPoint>>just(Collections.EMPTY_LIST);
+        }
+        syncDataPointsWithDataBase(apiDataPoints);
+        return Observable.just(apiDataPoints);
+    }
+
+    private void syncDataPointsWithDataBase(List<ApiDataPoint> apiDataPoints) {
+        if (apiDataPoints == null || apiDataPoints.size() == 0) {
+            return;
+        }
+        BriteDatabase.Transaction transaction = briteSurveyDbAdapter.beginTransaction();
+        try {
+            for (ApiDataPoint dataPoint : apiDataPoints) {
+                final String id = dataPoint.getId();
+                ContentValues values = new ContentValues();
+                values.put(RecordColumns.RECORD_ID, id);
+                values.put(RecordColumns.SURVEY_GROUP_ID, dataPoint.getSurveyGroupId());
+                values.put(RecordColumns.NAME, dataPoint.getDisplayName());
+                values.put(RecordColumns.LATITUDE, dataPoint.getLatitude());
+                values.put(RecordColumns.LONGITUDE, dataPoint.getLongitude());
+
+                syncSurveyInstances(dataPoint.getSurveyInstances(), id);
+
+                briteSurveyDbAdapter.updateRecord(id, values, dataPoint.getLastModified());
+            }
+            updateLastUpdatedDateTime(apiDataPoints);
+            transaction.markSuccessful();
+        } finally {
+            transaction.end();
+        }
+    }
+
+    /**
+     * JSON array responses are ordered to have the latest updated datapoint last so
+     * we record it to make the next query using it
+     * @param apiDataPoints
+     */
+    private void updateLastUpdatedDateTime(List<ApiDataPoint> apiDataPoints) {
+        ApiDataPoint surveyedLocale = apiDataPoints.get(apiDataPoints.size() - 1);
```

Comment:
I have added a @NonNull annotation and a check for empty list. If you look above it will not be called with empty or null but just in case.